README.md
# sta663-FinalProject-SGHMC
This repository houses the final project for STA 663 (Duke University Spring 2018), an implementation of stochastic gradient Hamiltonian Monte Carlo (SGHMC) as described by Chen et. al. in their 2014 paper "Stochastic Gradient Hamiltonian Monte Carlo".

The package associated with our work is housed in the SGHMC directory. The README file in the package documentation contains installation instructions and points the user to the examples. The following features are available in this package:
- an open source license (LICENSE.txt)
- Python and C++ implementations of the SGHMC algorithm (found in sghmc/sghmc_algorithm.py and sghmc/test/wrap.cpp, respectively)
- code to generate Examples 1 and 2 from the paper (found in sghmc/test/Example1.py and sghmc/test/Mixture_of_Normals.py, respectively)
- a Jupyter notebook containing the optimization work and results found in the optimization section of the final paper (found in sghmc/test/optimization_work.ipynb)
- a code to run all of the relevant examples (found in bin/run_examples.py)

The paper for our final project is housed in the paper directory, along with the figures generated in the package functions. 

The original paper upon which the project is based is available in the root folder of the Git repo.




SGHMC/CHANGES.txt
v<0.1.0>, <-4/25/2018> -- Initial release.


SGHMC/sghmc/test/wrap.cpp
<%
cfg['compiler_args'] = ['-std=c++11']
cfg['include_dirs'] = ['../../eigen3']
setup_pybind11(cfg)
%>

#include <pybind11/pybind11.h>
#include <pybind11/eigen.h>
#include <stdexcept>
#include <algorithm> // std::random_shuffle
#include <random>

#include <Eigen/LU>
#include <Eigen/Dense>

namespace py = pybind11;
using std::default_random_engine;
using std::normal_distribution;
        
// start random number engine with fixed seed
default_random_engine re{1234};
// set up random normal rnorm to work like in python
normal_distribution<double> norm(0, 1); // mean and standard deviation
auto rnorm = bind(norm, re);

// fill xs with draws from N(0,1) and return this n x 1 dim vector
Eigen::MatrixXd rnorm_vec(int n) {
    Eigen::MatrixXd xs = Eigen::MatrixXd::Zero(n, 1);
    for (int i=0; i<n; i++) {xs(i,0) = rnorm();}
    return xs;
}
    
// get noisy gradient of Fig1 example from Chen et. al. paper
Eigen::MatrixXd gradU_noisyFig1(Eigen::MatrixXd theta) {
    Eigen::MatrixXd xs = -4*theta.array() + 4*theta.array().pow(3) + 2*rnorm_vec(theta.rows()).array();
    return xs;
} 
    
// get gradient for mixture of normals example
Eigen::MatrixXd gradU_mixNormals(Eigen::MatrixXd theta, Eigen::MatrixXd x, int n, int batch_size) {
    int p = theta.rows();
    Eigen::ArrayXd c_0 = theta(0,0) - x.array();
    Eigen::ArrayXd c_1 = theta(1,0) - x.array();
    Eigen::ArrayXd star = 0.5 * (-0.5 * c_0.pow(2)).exp() + 0.5 * (-0.5 * c_1.pow(2)).exp();
    Eigen::ArrayXd star_prime;
    Eigen::MatrixXd grad = Eigen::MatrixXd::Zero(p, 1);
    for (int i=0; i<p; i++) {
        star_prime = 0.5 * (-0.5 * (theta(i,0) - x.array()).pow(2)).exp() * (theta(i,0) - x.array());
        grad(i,0) = -theta(i,0)/10 - (n/batch_size)*(star_prime/star).sum();
    }
    return grad;
} 

/* 
SGHMC algorithm as described in the paper by Chen et al.
Note that users specify the choice of which gradient function to use by gradU_choice, 
but the gradient function itself must be available in this file
*/
Eigen::MatrixXd sghmc(std::string gradU_choice, Eigen::MatrixXd eta, int niter, Eigen::MatrixXd alpha, Eigen::MatrixXd theta_0, Eigen::MatrixXd V_hat, Eigen::MatrixXd dat, int batch_size){
    // Initialization and checks
    int p = theta_0.rows(); // dimension of the thing you're sampling
    int n = dat.rows();     // number of data observations
    int p_dat = dat.cols(); // 2nd dimension of data
    int nbatches = n / batch_size; // how many batches data will be broken into
    // Set up dat_temp and dat_batch for use in loop below, as well as gradU_batch
    Eigen::MatrixXd dat_temp = dat;
    Eigen::MatrixXd dat_batch = Eigen::MatrixXd::Zero(batch_size, p_dat);
    Eigen::MatrixXd gradU_batch = Eigen::MatrixXd::Zero(p, 1);
    // set up matrix of 0s to hold samples
    Eigen::MatrixXd theta_samps = Eigen::MatrixXd::Zero(p, niter*(n/batch_size));
    // vector to hold indices for shuffling
    std::vector<int> ind;
    // fix beta_hat as described on pg. 6 of paper
    Eigen::MatrixXd beta_hat = 0.5 * V_hat * eta;
    // We're sampling from a N(0, 2(alpha - beta_hat) @ eta)
    // so this must be a positive definite matrix
    Eigen::MatrixXd Sigma = 2.0 * (alpha - beta_hat) * eta;
    Eigen::LLT<Eigen::MatrixXd> lltOfSig(Sigma); // compute the Cholesky decomposition of Sigma
    if(lltOfSig.info() == Eigen::NumericalIssue){ // check if we got error, and break out if so
        return theta_samps; // will just give back all-0s
    }
    Eigen::MatrixXd Sig_chol = lltOfSig.matrixL(); // get L in Chol decomp
    if(batch_size > n){ // Need batch size to be <= the amount of data
        return theta_samps; // will just give back all-0s
    }
    // initialize more things! (nu and theta, to be specific)
    Eigen::LLT<Eigen::MatrixXd> lltOfeta(eta); // compute the Cholesky decomposition of eta
    Eigen::MatrixXd eta_chol = lltOfeta.matrixL(); // get L in Chol decomp
    Eigen::MatrixXd nu = eta_chol * rnorm_vec(p); // initialize nu
    Eigen::MatrixXd theta = theta_0; // initialize theta 
    
    // loop through algorithm to get niter*batch_size samples
    int big_iter = 0;
    for (int it=0; it<niter; it++) {
        
        // shuffle rows of dat to get dat_temp 
        Eigen::VectorXi indices = Eigen::VectorXi::LinSpaced(dat.rows(), 0, dat.rows());
        std::random_shuffle(indices.data(), indices.data() + dat.rows());
        dat_temp = indices.asPermutation() * dat;
        
        // Resample momentum every epoch
        nu = eta_chol * rnorm_vec(p); // sample from MV normal
        
        // loop through the batches
        int count_lower = 0;
        int count_upper = batch_size;
        for (int b=0; b<nbatches; b++){
            int batch_ind = 0;
            for (int ind_temp=count_lower; ind_temp<count_upper; ind_temp++){
                dat_batch.row(batch_ind) = dat_temp.row(ind_temp);
                batch_ind += 1;
            }
            count_lower += batch_size; // add batch size to each iterator
            count_upper += batch_size;
            if (gradU_choice == "fig1"){
                gradU_batch = gradU_noisyFig1(theta);
            } else if (gradU_choice == "mixture_of_normals"){
                gradU_batch = gradU_mixNormals(theta, dat_batch, n, batch_size);
            } else {
                return theta_samps; // will just give back all-0s
            }
            nu = nu - eta * gradU_batch - alpha * nu + Sig_chol * rnorm_vec(p);
            theta = theta + nu;
            theta_samps.col(big_iter) = theta;
            big_iter += 1;
        }
    }

    return theta_samps;
}

PYBIND11_MODULE(wrap, m) {
    m.doc() = "auto-compiled c++ extension";
    m.def("sghmc", &sghmc);
}

SGHMC/sghmc/test/Example1.py
import matplotlib.pyplot as plt
import autograd.numpy as np
from autograd import jacobian
import seaborn as sns
import pystan
from sghmc.sghmc_algorithm import sghmc

### Easy test example based on Figure 1 from the paper
### See pg. 6 of Chen et. al. for their results/comparison

# Log likelihood function
def Ex1_sghmc():
    def U(theta):
        return(-2*theta**2 + theta**4)
    # True gradient
    gradU = jacobian(U, argnum=0)
    # Noisy gradient, based on what they do in the paper for Fig 1
    def noisy_gradU(theta, x, n, batch_size):
        '''Noisy gradient \Delta\tilde{U}(\theta)=\Delta U(\theta)+N(0,4)
        Extra args (x, n, batch_size) for compatibility with sghmc()'''
        return -4*theta + 4*theta**3 + np.random.normal(0,2)
    # Set random seed
    np.random.seed(1234)
    # Don't actually need 'data' in this example, just use
    # it as a place-holder to fit into our function.
    n = 100
    x = np.array([np.random.normal(0, 1, (n,1))]).reshape(-1,1)
    # Set up start values and tuning params
    theta_0 = np.array([0.0]) # Initialize theta
    p = theta_0.shape[0]
    eta = 0.001 * np.eye(p) # make this small
    alpha = 0.01 * np.eye(p)
    V = np.eye(p)
    batch_size = n # since we're not actually using the data, don't need to batch it
    niter = 500000 # Lots of iterations
    # run SGHMC sampler
    samps_sghmc = sghmc(noisy_gradU, eta, niter, alpha, theta_0, V, x, batch_size)
    
    # plot the samples from the algorithm and save to a file
    kdeplt = sns.kdeplot(samps_sghmc.reshape(-1)) # Plot the joint density
    fig = kdeplt.get_figure()
    fig.savefig('Example1_a.png')
    return(samps_sghmc)

################################################################################

# Use the hmc function from the pyhmc (at all default settings)
# to sample from this distribution.
# See https://pythonhosted.org/pyhmc/ for function details.
from pyhmc import hmc
def Ex1_hmc():
    def U(theta):
        return(-2*theta**2 + theta**4)
    # True gradient
    gradU = jacobian(U, argnum=0)
    # Noisy gradient, based on what they do in the paper for Fig 1
    def noisy_gradU(theta, x, n, batch_size):
        '''Noisy gradient \Delta\tilde{U}(\theta)=\Delta U(\theta)+N(0,4)
        Extra args (x, n, batch_size) for compatibility with sghmc()'''
        return -4*theta + 4*theta**3 + np.random.normal(0,2)
    # define your probability distribution
    def logprob(theta):
        logp = -2*theta**2 + theta**4
        grad = -4*theta + 4*theta**3
        return logp, grad
    # run the HMC sampler (use same theta_0 and niter as SGHMC)
    np.random.seed(1234)
    # Don't actually need 'data' in this example, just use
    # it as a place-holder to fit into our function.
    n = 100
    x = np.array([np.random.normal(0, 1, (n,1))]).reshape(-1,1)
    # Set up start values and tuning params
    theta_0 = np.array([0.0]) # Initialize theta
    batch_size = n # since we're not actually using the data, don't need to batch it
    niter = 500000 # Lots of iterations
    samps_hmc = hmc(logprob, x0=theta_0, n_samples=niter) 
    # plot the samples from the HMC algorithm
    sns.kdeplot(samps_hmc.reshape(-1))

    # The hmc() function with that many iterations goes kind of insane... 
    # How about less samples? Also redefine logprob(theta) to use funs from above
    #def logprob(theta):
        #return U(theta, x, n, batch_size), gradU(theta, x, n, batch_size).reshape(1)
    # run the HMC sampler (use same theta_0 as above but fewer samples)
   #samps_hmc = hmc(logprob, x0=theta_0, n_samples=500) # NOPE! Still looks bad.
    # plot the samples from the HMC algorithm
    
    # plot the samples from the algorithm and save to a file
    kdeplt = sns.kdeplot(samps_hmc.reshape(-1)) # Plot the joint density
    fig = kdeplt.get_figure()
    fig.savefig('Example1_b.png')
    
    return(samps_hmc)

################################################################################

### Make figure showing the truth compared to our samples
def Ex1_truth():
    # define exp(-U(theta)), the true density
    def expU(theta):
        return np.exp(-1*U(theta))/5.36516 # norm constant from Wolfram Alpha

    samps_sghmc = np.load("samps_sghmc.npy") # load data run with tons of iters from C++ code
    test_theta = np.linspace(samps_sghmc.min(),samps_sghmc.max(),200)
    
    # just truth
    fig, ax = plt.subplots( nrows=1, ncols=1 )  # create figure & 1 axis
    ax.plot(test_theta,expU(test_theta))
    fig.savefig('Example1_truth.png')
    
    # truth and C++ samples
    #fig, ax = plt.subplots()
    #sns.kdeplot(samps_sghmc[0,:], ax=ax)  # plot samples (blue)
    #plt.plot(test_theta,expU(test_theta)) # plot true density (orange)

################################################################################

# Stan code for our Example 1 :
import pystan
def Ex1_stan():
    stan_ex1_code = '''
    functions {
      // Define log probability density function
      real Ex1_lpdf(real theta) {return -1* (-2*theta^2 + theta^4);}
    }
    data {
    }
    parameters {
      real theta;
    }
    model {
      // Sample theta
      theta ~ Ex1_lpdf();
    }
    
    '''
    ex1_dat = {}
    
    sm = pystan.StanModel(model_code=stan_ex1_code)
    fit = sm.sampling(data=ex1_dat, iter=100000, chains=4)
    
    # return a dictionary of arrays
    la = fit.extract(permuted=True)  # return a dictionary of arrays
    thet_samps = la['theta']
    
    # PLOT and save output
    kdeplt = sns.kdeplot(thet_samps)
    fig = kdeplt.get_figure()
    fig.savefig('Example1_c.png')
    
    samps_stan = thet_samps
    return(samps_stan)


SGHMC/sghmc/test/__init__.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Wed Apr 25 10:44:16 2018

@author: isaaclavine
"""



examples/fig3.py
### From their Figure 3 example at these links:
#https://github.com/tqchen/ML-SGHMC/blob/master/matlab/figure3/sghmc.m
#https://github.com/tqchen/ML-SGHMC/blob/master/matlab/figure3/figure3b.m

# WHY DOESN'T THIS WORK???

import matplotlib.pyplot as plt
import autograd.numpy as np
from autograd import jacobian
import seaborn as sns

def sghmc( gradU, eta, L, alpha, x, V ):
    m = len(x)
    samps = np.zeros( (m, L) )
    beta = V * eta * 0.5

    if beta > alpha:
        print('too big eta');

    sigma = np.sqrt( 2 * eta * (alpha-beta) )
    p = np.random.normal(size=( m, 1 )) * np.sqrt( eta )
    momentum = 1 - alpha

    for t in range(L):
        p = p * momentum - gradU( x ) * eta + np.random.normal(size=(2,1)) * sigma
        x = x + p
        samps[:,t] = x.reshape(-1,2)
    
    return samps

V = 1
# covariance matrix
rho = 0.9
covS = np.array([ [1, rho], [rho, 1] ])
invS = np.linalg.inv(covS)
# intial x
x = np.zeros(2).reshape(2,1)
alpha = 0.05
etaSGHMC = 0.05
# number of steps 
L = 1000
nset = 5

def log_norm(X):
    return(-(1/2)*np.log(np.linalg.det(covS))-(1/2)*X.T@invS@X)

# Automatic differentiation to get the jacobian
gradUTrue = jacobian(log_norm, argnum=0)

# make sure grad gives what it should
print(-invS@np.array([1.0,1.0]))
print(gradUTrue(np.array([1.0,1.0])))

# define noisy gradient like they do
def gradUNoise(X):
    return -invS@X  + np.random.normal(size=( 2, 1 ))

'''
for i = 1 : nset
    dscale = (0.6^(i-1));
    eta = etaSGHMC * dscale*dscale;
    dsghmc = sghmc( gradUNoise, eta, L, alpha*dscale, x, V );
    covESGHMC(:,:,i) = dsghmc * dsghmc' / L;
    meanESGHMC(:,i) = mean( dsghmc, 2 );
    SGHMCeta(i) = eta;
    SGHMCauc(i) = mean(aucTime( dsghmc, 1 ));
'''

i = 3
dscale = (0.6**(i-1))
eta = etaSGHMC * dscale*dscale
dsghmc = sghmc( gradUNoise, eta, L, alpha*dscale, x, V )

# look at first dimension (should be normal)
sns.kdeplot(dsghmc[0,:])

SGHMC/sghmc/sghmc_algorithm.py
import autograd.numpy as np


def is_pos_def(X):
    '''Check whether a matrix X is pos definite.
    Returns True or False, depending on X.
    '''
    return np.all(np.linalg.eigvals(X) > 0)

def batch_data(data, batch_size):
    n = data.shape[0]
    p = data.shape[1]
    if n % batch_size != 0:
        n = (n // batch_size) * batch_size
    ind = np.arange(n)
    np.random.shuffle(ind)
    n_batches = n // batch_size
    data = data[ind].reshape(batch_size, p, n_batches)
    return(data, n_batches)

def sghmc(gradU, eta, niter, alpha, theta_0, V_hat, dat, batch_size):
    '''Define SGHMC as described in the paper
    Tianqi Chen, Emily B. Fox, Carlos Guestrin 
    Stochastic Gradient Hamiltonian Monte Carlo 
    ICML 2014.

    The inputs are:
    gradU = gradient of U
    eta = eps^2 M^(-1)
    niter = number of samples to generate
    alpha = eps M^(-1) C
    theta_0 = initial val of parameter(s) to be sampled
    V_hat = estimated covariance matrix of stoch grad noise
    See paper for more details

    The return is:
    A np.array of positions of theta.'''

    ### Initialization and checks ###
    # get dimension of the thing you're sampling
    p = len(theta_0)
    # set up matrix of 0s to hold samples
    n = dat.shape[0]
    theta_samps = np.zeros((p, niter*(n // batch_size)))
    # fix beta_hat as described on pg. 6 of paper
    beta_hat = 0.5 * V_hat @ eta
    # We're sampling from a N(0, 2(alpha - beta_hat) @ eta)
    # so this must be a positive definite matrix
    Sigma = 2 * (alpha - beta_hat) @ eta
    if not is_pos_def( Sigma ): 
        print("Error: (alpha - beta_hat) eta not pos def")
        return
    # Need batch size to be <= the amount of data
    if (batch_size > n): 
        print("Error: batch_size must be <= number of data points")
        return

    # initialize nu and theta 
    nu = np.random.multivariate_normal(np.zeros(p), eta).reshape(p,-1)
    theta = theta_0
    
    # loop through algorithm to get niter samples
    it = 0
    for i in range(niter):
        dat_resh, nbatches = batch_data(dat, batch_size)
        
        # Resample momentum every epoch
        nu = np.random.multivariate_normal(np.zeros(p), eta).reshape(p,-1)
        
        for batch in range(nbatches):
            gradU_batch = gradU(theta, dat_resh[:,:,batch], n, batch_size).reshape(p,-1)
            nu = nu - eta @ gradU_batch - alpha @ nu + \
                 np.random.multivariate_normal(np.zeros(p), Sigma).reshape(p, -1)
            theta = theta + nu
            theta_samps[:,it] = theta.reshape(-1,p)
            it = it + 1
        
    return theta_samps

def sghmc_cleaned(gradU, eta, niter, alpha, theta_0, V_hat, dat, batch_size):
    '''Define SGHMC as described in the paper
    Tianqi Chen, Emily B. Fox, Carlos Guestrin 
    Stochastic Gradient Hamiltonian Monte Carlo 
    ICML 2014.

    The inputs are:
    gradU = gradient of U
    eta = eps^2 M^(-1)
    niter = number of samples to generate
    alpha = eps M^(-1) C
    theta_0 = initial val of parameter(s) to be sampled
    V_hat = estimated covariance matrix of stoch grad noise
    See paper for more details

    The return is:
    A np.array of positions of theta.'''

    ### Initialization and checks ###
    # get dimension of the thing you're sampling
    p = len(theta_0)
    # set up matrix of 0s to hold samples
    n = dat.shape[0]
    theta_samps = np.zeros((p, niter*(n // batch_size)))
    # fix beta_hat as described on pg. 6 of paper
    beta_hat = 0.5 * V_hat @ eta
    # We're sampling from a N(0, 2(alpha - beta_hat) @ eta)
    # so this must be a positive definite matrix
    Sigma = 2 * (alpha - beta_hat) @ eta
    Sig_chol = np.linalg.cholesky(Sigma)
    if not is_pos_def( Sigma ): 
        print("Error: (alpha - beta_hat) eta not pos def")
        return
    # Need batch size to be <= the amount of data
    if (batch_size > n): 
        print("Error: batch_size must be <= number of data points")
        return

    # initialize nu and theta 
    nu = np.random.multivariate_normal(np.zeros(p), eta).reshape(p,-1)
    theta = theta_0

    # set up for Chol decomp for MV normal sampling of nu every epoch
    eta_chol = np.linalg.cholesky(eta)

    # loop through algorithm to get niter samples
    it = 0
    for i in range(niter):
        dat_resh, nbatches = batch_data(dat, batch_size)
        
        # Resample momentum every epoch
        nu = eta_chol @ np.random.normal(size=p).reshape(p,-1) # sample from MV normal
        
        for batch in range(nbatches):
            gradU_batch = gradU(theta, dat_resh[:,:,batch], n, batch_size).reshape(p,-1)
            nu = nu - eta @ gradU_batch - alpha @ nu + \
                 Sig_chol @ np.random.normal(size=p).reshape(p,-1) # sample from MV normal
            theta = theta + nu
            theta_samps[:,it] = theta.reshape(-1,p)
            it = it + 1
        
    return theta_samps

examples/Mixture_of_Normals.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Thu Apr 19 15:36:51 2018

@author: isaaclavine
"""

import matplotlib.pyplot as plt
import autograd.numpy as np
from autograd import jacobian
import seaborn as sns

## Example #1:
## Sampling from a mixture of normals in 1-D
## SAMPLING MODEL: x ~ 0.5 * N(mu1, 1) + 0.5 * N(mu2, 1)
## PRIORS: p(mu1) = p(mu2) = N(0,10)

def log_prior(theta):
    return(-(1/(2*10))*theta.T@theta)
      
def log_lik(theta, x):
    return(np.log(0.5 * np.exp(-0.5*(theta[0]-x)**2) + 0.5* np.exp(-0.5*(theta[1]-x)**2)))

def U(theta, x, n, batch_size):
    return(-log_prior(theta) - (n/batch_size)*sum(log_lik(theta, x)))
       
#def U2(theta, **kwargs):
#    return(-log_prior(theta) - (n/batch_size)*sum(log_lik(theta, x)))
    
#kwargs = {"n":n, "batch_size":batch_size, "x":x}

#U2(theta, **kwargs)
    
# Automatic differentiation to get the gradient
gradU = jacobian(U, argnum=0)

# Set random seed
np.random.seed(1234)
# Set up the data
p = 2 #dimension of theta
theta = np.array([-3.0, 3.0]).reshape(p,-1)
n = 10000
x = np.array([np.random.normal(theta[0], 1, (n,1)),
              np.random.normal(theta[1], 1, (n,1))]).reshape(-1,1)

# Plot the empirical distribution of x (sampled from mixture of normals)
sns.kdeplot(x.reshape(-1), bw=1)

## Initialize parameters and sample 

# Initialize mean parameters
#theta_0 = np.random.normal(size=(p,1))
theta_0 = theta # initialize at "true" value for testing

# Initialize tuning parameters:
# learning rate
eta = 0.01/n * np.eye(p)
# Friction rate
alpha = 0.1 * np.eye(p)

# Arbitrary guess at covariance of noise from mini-batching the data
V = np.eye(p)*1
niter = 100
batch_size=1000

# Run sampling algorithm
samps = sghmc(gradU, eta, niter, alpha, theta_0, V, x, batch_size)

# See below that the mean estimates are VERY precice
# using this sampler, e.g. for \mu_0 we have samples in the range of
# roughly 2.96 to 3.04, and for \mu_1 from -3.05 to -2.98 

# Plot the density of each mean in mixture
sns.kdeplot(samps[0,:]) # MLE is -3, prior shrinks this to 0 a little
sns.kdeplot(samps[1,:]) # MLE is 3, prior shrinks this to 0 a little

# Plot the joint density
sns.kdeplot(samps[0,:], samps[1,:])


################################################################################

# Use the hmc function from the pyhmc (at all default settings)
# to sample from this distribution.
# See https://pythonhosted.org/pyhmc/ for function details.
from pyhmc import hmc

# define your probability distribution
# note some fiddling to get dimensions to be compatible
def logprob(theta):
    logp = np.sum(U(theta, x=x, n=n, batch_size=n))
    gradu = gradU(theta, x=x, n=n, batch_size=n).reshape((-1,))
    return logp, gradu

# run the HMC sampler 
# ideally would use same theta_0 and niter as SGHMC, 
# but computing the full gradient is prohibtively slow!!
samps_hmc = hmc(logprob, x0=theta_0.reshape((-1)), n_samples=100) 
# plot the samples from the HMC algorithm (jointly)
sns.kdeplot(samps[0,:], samps[1,:])


examples/Fig1_toy_example.py
import matplotlib.pyplot as plt
import autograd.numpy as np
from autograd import jacobian
import seaborn as sns

### Easy test example based on Figure 1 from the paper
### See pg. 6 of Chen et. al. for their results/comparison

# Log likelihood function
def U(theta):
    return(-2*theta**2 + theta**4)
# True gradient
gradU = jacobian(U, argnum=0)
# Noisy gradient, based on what they do in the paper for Fig 1
def noisy_gradU(theta, x, n, batch_size):
    '''Noisy gradient \Delta\tilde{U}(\theta)=\Delta U(\theta)+N(0,4)
    Extra args (x, n, batch_size) for compatibility with sghmc()'''
    return -4*theta + 4*theta**3 + np.random.normal(0,2)
# Set random seed
np.random.seed(1234)
# Don't actually need 'data' in this example, just use
# it as a place-holder to fit into our function.
n = 100
x = np.array([np.random.normal(0, 1, (n,1))]).reshape(-1,1)
# Set up start values and tuning params
theta_0 = np.array([0.0]) # Initialize theta
p = theta_0.shape[0]
eta = 0.001 * np.eye(p) # make this small
alpha = 0.01 * np.eye(p)
V = np.eye(p)
batch_size = n # since we're not actually using the data, don't need to batch it
niter = 500000 # Lots of iterations
# run SGHMC sampler
samps_sghmc = sghmc(noisy_gradU, eta, niter, alpha, theta_0, V, x, batch_size)
# plot the samples from the algorithm
sns.kdeplot(samps_sghmc.reshape(-1))

################################################################################

# Use the hmc function from the pyhmc (at all default settings)
# to sample from this distribution.
# See https://pythonhosted.org/pyhmc/ for function details.
from pyhmc import hmc
# define your probability distribution
def logprob(theta):
    logp = -2*theta**2 + theta**4
    grad = -4*theta + 4*theta**3
    return logp, grad
# run the HMC sampler (use same theta_0 and niter as SGHMC)
samps_hmc = hmc(logprob, x0=theta_0, n_samples=niter) 
# plot the samples from the HMC algorithm
sns.kdeplot(samps_hmc.reshape(-1))

# The hmc() function with that many iterations goes kind of insane... 
# How about less samples? Also redefine logprob(theta) to use funs from above
def logprob(theta):
    return U(theta, x, n, batch_size), gradU(theta, x, n, batch_size).reshape(1)
# run the HMC sampler (use same theta_0 as above but fewer samples)
samps_hmc = hmc(logprob, x0=theta_0, n_samples=500) # NOPE! Still looks bad.
# plot the samples from the HMC algorithm
sns.kdeplot(samps_hmc.reshape(-1))
